### PR TITLE
Linux Exception safe

### PR DESCRIPTION
**Pull request by AStueben**
# Linux exception safe layer 
As mentioned in https://github.com/shimat/opencvsharp/issues/517, exception happening on opencv-side causes the program to crash on Linux. An idea for a fix would be to introduce and additional layer both on the C++ and the C# side to catch an occurring exception and and handle it on the native side as sketched out here: https://github.com/shimat/opencvsharp/issues/517#issuecomment-411477907. 
This pull request shows a possible (but unfinished) implementation using reflection to generate the code on C# side and a python script generating the code on C++ side by scanning the files in OpenCvSharpExtern.

### CODE REVIEW

**Comment by AStueben** [comment]
The python script lies in ```src/OpenCvSharpCodeGen/cpp_gen/generate_header_file.py``` and does not require any modification to the current project. 
The C# project in  ```src/OpenCvSharpCodeGen/csharp_gen/```  needs a functioning NativeMethod class to create both new outer layer (replacements for the existing ```[DllImport]``` methods) and the inner layer (the slightly altered signatures for the generated C++ methods).

Until no I have yet to find a solution to instantiate the NativeMethods class separately from the project; a.t.m. to run the C# generator one must 
* rename the ```src/OpenCvSharp/.PInvoke``` folder to ```PInvoke``` and the ```PInvokeNew``` to ```.PInvokeNew``` 
* run the project in ```src/OpenCvSharpCodeGen/csharp_gen/```
* rename the ```PInvoke``` back to ```.PInvoke``` and the ```.PInvokeNew``` back to ```PInvokeNew``` 
* move both files ```NativeMethods``` and ```NativeMethodsExc```in ```src/OpenCvSharpCodeGen/``` to  the ```PInvokeNew```

File: src/OpenCvSharp/PInvokeNew/ExceptionHandler.cs [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+        private static ThreadLocal<OpenCvSharp.ErrorCode> localStatus = new ThreadLocal<OpenCvSharp.ErrorCode>();
+        private static ThreadLocal<string> localFuncName    = new ThreadLocal<string>();
+        private static ThreadLocal<string> localErrMsg      = new ThreadLocal<string>();
+        private static ThreadLocal<string> localFileName    = new ThreadLocal<string>();
+        private static ThreadLocal<int> localLine           = new ThreadLocal<int>();
+
+        /// <summary>
+        /// Registers the callback function to opencv, so exception caught befor the p/invoke boundary 
+        /// </summary>
+        public static void registerExceptionCallback()
```
Comment (shimat) [inline src/OpenCvSharp/PInvokeNew/ExceptionHandler.cs]: After merging, I will modify method names to PascalCase

**Comment by shimat** [comment]
Thank you. I am sorry but I still don't really understand details of your work very well. I am reading the diffs.
This is a helpful change for Linux, but for Windows (.NET Framework) this patch is unnecessary and slows down all P/Invoke calls. I am worried about whether to add conditional branchings by platform or OS.

**Comment by AStueben** [comment]
Thank you for your input. I tried to be as non-invasive to the original code as I could. 

As with my last commits to this pull request, I introduced an idea which leaves the original code nearly untouched (some files are added/moved), so that under Windows the invocation of the native code uses the same classes as before. 

Under Linux, using the bash script [OpenCvSharpCodeGen/generate_exc_safe.sh](https://github.com/Xamla/opencvsharp/blob/e96e3fac2d0b9354ecf5cb92a449b8205db2dcf5/src/OpenCvSharpCodeGen/generate_exc_safe.sh), one can generate the appropriate files under Linux.
After that, the cmake build process can be started.  
What the script does, beside generating the C# (using reflection of the functional OpenCvSharp-project) and C++ files and moving them into OpenCvSharp(Extern), is renaming the folder [src/OpenCvSharp/PInvoke/NativeMethods](https://github.com/Xamla/opencvsharp/tree/e96e3fac2d0b9354ecf5cb92a449b8205db2dcf5/src/OpenCvSharp/PInvoke/NativeMethods) to .NativeMethods (hidden) and adding a new NativeMethods folder, which now contains the C# files replacing the ```[DllImports]``` calls and adding a NativeMethodsExc class, which implements the new ```[DllImports]```.

The script [OpenCvSharpCodeGen/reset_exc_safe.sh](https://github.com/Xamla/opencvsharp/blob/e96e3fac2d0b9354ecf5cb92a449b8205db2dcf5/src/OpenCvSharpCodeGen/reset_exc_safe.sh) can be used to reverse the effect.

This approach introduces some dependencies, since every file containing ```[DllImports]``` methods of the partial class ```NativeMethods``` must reside in  [src/OpenCvSharp/PInvoke/NativeMethods](https://github.com/Xamla/opencvsharp/tree/e96e3fac2d0b9354ecf5cb92a449b8205db2dcf5/src/OpenCvSharp/PInvoke/NativeMethods). 

Please let me know what you think of this approach, if it is an acceptable solution and (if not) what could be improved.

File: src/OpenCvSharpCodeGen/csharp_gen/Helper.cs [hunk 2]
```diff
@@ -0,0 +1,52 @@
+using System;
+using System.Collections;
+
+namespace ExceptionSafeGenerator
+{
+    static class Helper
+    {
+        static public string CommaSeparated(IEnumerable list)
+        {
+            string outString = "";
```
Comment (shimat) [inline src/OpenCvSharpCodeGen/csharp_gen/Helper.cs:10]: This method can be replaced by string.Join
```c#
string.Join(“,”, list);
```

**Comment by shimat** [comment]
Thank you for your waiting. I understand and accept your approach :+1: 

I usually do opencvsharp development only for Windows. So, maintenance of this work is an issue. One solution is to use a CI system (to change AppVeyor settings to support Linux), but to be honest, I cannot spend much time to do that so far. I understand that I should do it 🤔

**Comment by turowicz** [comment]
+1 from me

**Comment by shimat** [comment]
Fixed by #758. thanks!

**Pull request closed without merging**